crates/key-derivation/src/operator.rs

```diff
@@ -0,0 +1,453 @@
+        }
+        master.private_key.non_secure_erase();
+        {
+            let chaincode: &mut [u8; 32] = master.chain_code.as_mut();
+            chaincode.zeroize();
+        }
+        let raw_ptr = &mut master.child_number as *mut ChildNumber;
+        // SAFETY: `master.child_number` is a valid enum variant
+        //          and will not be accessed after zeroization.
+        //          Also there are only two possible variants that will
+        //          always have an `index` which is a `u32`.
+        //          Note that `ChildNumber` does not have the `#[non_exhaustive]`
+        //          attribute.
+        unsafe {
+            *raw_ptr = if master.child_number.is_normal() {
+                ChildNumber::Normal { index: 0 }
+            } else {
+                ChildNumber::Hardened { index: 0 }
+            };
+        }
+
+        // Zeroize base components
+        base.depth.zeroize();
+        {
+            let fingerprint: &mut [u8; 4] = base.parent_fingerprint.as_mut();
+            fingerprint.zeroize();
+        }
+        base.private_key.non_secure_erase();
+        {
+            let chaincode: &mut [u8; 32] = base.chain_code.as_mut();
+            chaincode.zeroize();
+        }
+        let raw_ptr = &mut base.child_number as *mut ChildNumber;
+        // SAFETY: `base.child_number` is a valid enum variant
+        //          and will not be accessed after zeroization.
+        //          Also there are only two possible variants that will
+        //          always have an `index` which is a `u32`.
+        //          Note that `ChildNumber` does not have the `#[non_exhaustive]`
+        //          attribute.
+        unsafe {
+            *raw_ptr = if base.child_number.is_normal() {
+                ChildNumber::Normal { index: 0 }
+            } else {
+                ChildNumber::Hardened { index: 0 }
+            };
+        }
+
+        // Zeroize message components
+        message.depth.zeroize();
+        {
+            let fingerprint: &mut [u8; 4] = message.parent_fingerprint.as_mut();
+            fingerprint.zeroize();
+        }
+        message.private_key.non_secure_erase();
+        {
+            let chaincode: &mut [u8; 32] = message.chain_code.as_mut();
+            chaincode.zeroize();
+        }
+        let raw_ptr = &mut message.child_number as *mut ChildNumber;
+        // SAFETY: `message.child_number` is a valid enum variant
+        //          and will not be accessed after zeroization.
+        //          Also there are only two possible variants that will
+        //          always have an `index` which is a `u32`.
+        //          Note that `ChildNumber` does not have the `#[non_exhaustive]`
+        //          attribute.
+        unsafe {
+            *raw_ptr = if message.child_number.is_normal() {
+                ChildNumber::Normal { index: 0 }
+            } else {
+                ChildNumber::Hardened { index: 0 }
+            };
+        }
```

Comment:
In theory the compiler could optimize this in the absence of an optimization barrier. But the consequences of this seem pretty trivial.

Reply:
Yes, but even with the `#[inline]`?